File: src/Day1.hs
module Day1 where

import Data.List as List
import Data.Maybe as M
import Data.Set as Set

main :: IO ()
main = do
    numbers <- fmap read . lines <$> readFile "puzzle-inputs/1.txt"
    putStrLn $
        unlines
            [ "Day 1"
            , case twoNumbersThatAddTo 2020 numbers of
                  Nothing -> "No two numbers add to 2020"
                  Just (one , two) ->
                      unwords [show one , "and" , show two , "which multiply to" , show (one * two)]
            , case threeNumbersThatAddTo 2020 numbers of
                  Nothing -> "No three numbers add to 2020"
                  Just (one , two , three) ->
                      unwords [show one , "," , show two , ", and" , show three , "which multiply to" , show (one * two * three)]
            ]

-- O(n)
twoNumbersThatAddTo :: Int -> [Int] -> Maybe (Int , Int)
twoNumbersThatAddTo target numbers = M.listToMaybe $ do
    let nset = Set.fromList numbers
    n <- numbers
    let complement = target - n
    if complement `Set.member` nset
        then return (n , complement)
        else mempty

-- O(n^2)
threeNumbersThatAddTo :: Int -> [Int] -> Maybe (Int , Int , Int)
threeNumbersThatAddTo target numbers = listToMaybe $ do
    x <- numbers
    Just (y , z) <- pure $ twoNumbersThatAddTo (target - x) $ List.filter (/= x) numbers
    return (x , y , z)

File: src/day1.cpp
#include <iostream>
#include <fstream>
#include <set>
#include <tuple>
#include "maybe.cpp"

using namespace std;

Maybe<tuple<int, int>> two_add_to(int n, set<int> &xs) {
    for (int i : xs) {
        int match = n - i;
        if (xs.count(match)) {
            cout << i << " : " << match << "\n";
            return Maybe<tuple<int, int>>({i, match});
        }
    }
    return Nothing<tuple<int, int>>();
}

Maybe<tuple<int, int, int>> three_add_to(int n, set<int> &xs) {
    for (int i : xs) {
        auto two = two_add_to(n - i, xs);
        auto nums = two.withDefault({i, i});
        if (get<0>(nums) != i)
            return Just<tuple<int, int, int>>({i, get<0>(nums), get<1>(nums)});
    }
    return Nothing<tuple<int, int, int>>();
}

int main () {
    ifstream file ("./puzzle-inputs/1.txt");
    set<int> numbers;

    {   // Add all numbers to numbers;
        string line;
        while (getline(file, line)) numbers.insert(stoi(line));
    }

    {
        auto t = two_add_to(2020, numbers);
        t.run<void>(
            [](tuple<int, int> v){
                cout
                    << "Result: "
                    << get<0>(v)
                    << " and "
                    << get<1>(v)
                    << " which add to "
                    << get<0>(v) * get<1>(v)
                    << "\n";
            },
            [](void) {
                cout << "No result found";
            }
        );
    }

    {
        auto t = three_add_to(2020, numbers);
        t.run<void>(
            [](tuple<int, int, int> v){
                cout
                    << "Result: "
                    << get<0>(v)
                    << ", "
                    << get<1>(v)
                    << " and "
                    << get<1>(v)
                    << " which add to "
                    << get<0>(v) * get<1>(v) * get<2>(v)
                    << "\n";
            },
            [](void) {
                cout << "No result found";
            }
        );
    }

    return 0;
}
